src/log_batch.rs

```diff
@@ -79,92 +80,92 @@ impl<M: MessageExt> Entries<M> {
         }
     }
 
-    pub fn from_bytes(
-        buf: &mut SliceReader<'_>,
-        file_id: FileId,
-        base_offset: u64,  // Offset of the batch from its log file.
-        batch_offset: u64, // Offset of the item from in its batch.
-        entries_size: &mut usize,
-    ) -> Result<Entries<M>> {
-        let content_len = buf.len() as u64;
-        let mut count = codec::decode_var_u64(buf)? as usize;
-        trace!("decoding entries, count: {}", count);
-        let mut entries = Vec::with_capacity(count);
-        let mut entries_index = Vec::with_capacity(count);
+    pub fn from_bytes(buf: &mut SliceReader<'_>, entries_size: &mut usize) -> Result<Entries<M>> {
+        let mut count = codec::decode_var_u64(buf)?;
+        let mut entries_index = Vec::with_capacity(count as usize);
         while count > 0 {
-            let len = codec::decode_var_u64(buf)? as usize;
-            trace!("decoding an entry, len: {}", len);
-            let mut e = M::Entry::new();
-            e.merge_from_bytes(&buf[..len])?;
-
+            let index = codec::decode_var_u64(buf)?;
+            let len = codec::decode_var_u64(buf)? - *entries_size as u64;
             let entry_index = EntryIndex {
-                index: M::index(&e),
-                file_id,
-                base_offset,
-                offset: batch_offset + content_len - buf.len() as u64,
-                len: len as u64,
+                index,
+                offset: *entries_size as u64,
+                len,
                 ..Default::default()
             };
-            *entries_size += entry_index.len as usize;
-
-            buf.consume(len);
-            entries.push(e);
+            *entries_size += len as usize;
             entries_index.push(entry_index);
-
             count -= 1;
         }
-        Ok(Entries::new(entries, Some(entries_index)))
+        Ok(Entries::new(vec![], Some(entries_index)))
     }
 
-    pub fn encode_to(&mut self, vec: &mut Vec<u8>, entries_size: &mut usize) -> Result<()> {
-        // layout = { entries count | multiple entries }
-        // entries layout = { entry layout | ... | entry layout }
-        // entry layout = { len | entry content }
-        vec.encode_var_u64(self.entries.len() as u64)?;
+    pub fn encode_to(
+        &mut self,
+        buf: &mut Vec<u8>,
+        entries_buf: &mut Vec<u8>,
+        entries_size: &mut usize,
+    ) -> Result<()> {
+        // buf layout = { entries count | [ entry_index | entry_tail_offset ] }
```

Comment:
Only need to encode the first index, indexes in one `Entries` are always contiguous

Reply:
Ohh, I forgot that. Let me add an `assert` here to make sure of it.